src/ralph/backends/data/base.py

```diff
@@ -0,0 +1,161 @@
+class DataBackendStatus(Enum):
+    """Data backend statuses."""
+
+    OK = "ok"
+    AWAY = "away"
+    ERROR = "error"
+
+
+def enforce_query_checks(method):
+    """Enforces query argument type checking for methods using it."""
+
+    @functools.wraps(method)
+    def wrapper(*args, **kwargs):
+        """Wrap method execution."""
+        query = kwargs.pop("query", None)
+        self_ = args[0]
+
+        return method(*args, query=self_.validate_query(query), **kwargs)
+
+    return wrapper
+
+
+class BaseDataBackend(ABC):
+    """Base data backend interface."""
+
+    name = "base"
+    query_model = BaseQuery
+    default_operation_type = BaseOperationType.INDEX
+    settings_class = BaseDataBackendSettings
+
+    @abstractmethod
+    def __init__(self, settings: settings_class = None):
+        """Instantiates the data backend."""
+
+    def validate_query(self, query: Union[str, dict, BaseQuery] = None) -> BaseQuery:
+        """Validates and transforms the query."""
+        if query is None:
+            query = self.query_model()
+
+        if isinstance(query, str):
+            query = self.query_model(query_string=query)
+
+        if isinstance(query, dict):
+            try:
+                query = self.query_model(**query)
+            except ValidationError as err:
+                raise BackendParameterException(
+                    "The 'query' argument is expected to be a "
+                    f"{self.query_model().__class__.__name__} instance. {err.errors()}"
+                ) from err
+
+        if not isinstance(query, self.query_model):
+            raise BackendParameterException(
+                "The 'query' argument is expected to be a "
+                f"{self.query_model().__class__.__name__} instance."
+            )
+
+        logger.debug("Query: %s", str(query))
+
+        return query
+
+    @abstractmethod
+    def status(self) -> DataBackendStatus:
+        """Implements data backend checks (e.g. connection, cluster status)."""
+
+    @abstractmethod
+    def list(
+        self, target: str = None, details: bool = False, new: bool = False
+    ) -> Iterator[Union[str, dict]]:
+        """Lists containers in the data backend. E.g., collections, files, indexes.
+
+        Yields:
+            str: If `details` is False.
+            dict: If `details` is True.
+        """
+
+    @abstractmethod
+    @enforce_query_checks
+    def read(
+        self,
+        *,
+        query: Union[str, BaseQuery] = None,
+        target: str = None,
+        chunk_size: Union[None, int] = None,
+        raw_output: bool = False,
+    ) -> Iterator[Union[bytes, dict]]:
+        """Reads records matching the query in the target container and yields them.
+
+        Yields:
+            dict: If `raw_output` is False.
+            bytes: If `raw_output` is True.
+        """
+
+    @abstractmethod
+    def write(  # pylint: disable=too-many-arguments
+        self,
+        data: Iterable[Union[bytes, dict]],
```

Comment:
Quick question: with `data` being an Iterable, are we losing, using Ralph as the CLI, the ability to stream from a Linux pipe to write on a backend?

Reply:
That's a good point) file-like objects like standard IO streams we use in the CLI are Iterable thus, this should be covered, however, it might indeed be useful to explicitly specify IO streams in the annotation to highlight this use case) Will update that)